pkg/catalogv2/git/git.go

```diff
@@ -1,480 +1,264 @@
-func (g *git) Clone(branch string) error {
-	if branch == "" {
-		return g.git("clone", "--depth=1", "-n", "--", g.URL, g.Directory)
-	}
-	return g.git("clone", "--depth=1", "-n", "--branch="+branch, "--", g.URL, g.Directory)
+	ro.localRepo = localRepository
+	return nil
 }
 
-// Update updates git repo if remote sha has changed
-func (g *git) Update(branch string) (string, error) {
-	if err := g.clone(branch); err != nil {
-		return "", nil
-	}
-
-	if err := g.reset("HEAD"); err != nil {
-		return "", err
-	}
-
-	commit, err := g.currentCommit()
+// getCurrentCommit returns the commit hash of the HEAD from the current branch
+func (ro *repoOperation) getCurrentCommit() (plumbing.Hash, error) {
+	headRef, err := ro.localRepo.Head()
 	if err != nil {
-		return commit, err
-	}
-
-	if changed, err := g.remoteSHAChanged(branch, commit); err != nil || !changed {
-		return commit, err
-	}
-
-	if err := g.fetchAndReset(branch); err != nil {
-		return "", err
+		return plumbing.Hash{}, fmt.Errorf("getCurrentCommit failure: %w", err)
 	}
 
-	return g.currentCommit()
+	return headRef.Hash(), nil
 }
 
-// Ensure runs git clone, clean DIRTY contents and fetch the latest commit
-func (g *git) Ensure(commit string) error {
-	if err := g.clone(""); err != nil {
-		return err
-	}
+func (ro *repoOperation) updateRefSpec(branch string) {
+	var newRefSpec string
 
-	if err := g.reset(commit); err == nil {
-		return nil
+	if branch != "" {
+		newRefSpec = fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", branch, branch)
+	} else {
+		newRefSpec = "+refs/heads/*:refs/remotes/origin/*"
 	}
 
-	return g.fetchAndReset(commit)
+	if len(ro.fetchOpts.RefSpecs) > 0 {
+		ro.fetchOpts.RefSpecs[0] = config.RefSpec(newRefSpec)
+	} else {
+		ro.fetchOpts.RefSpecs = []config.RefSpec{config.RefSpec(newRefSpec)}
+	}
 }
 
-func (g *git) httpClientWithCreds() (*http.Client, error) {
-	var (
-		username  string
-		password  string
-		tlsConfig tls.Config
-	)
-
-	if g.secret != nil {
-		switch g.secret.Type {
-		case corev1.SecretTypeBasicAuth:
-			username = string(g.secret.Data[corev1.BasicAuthUsernameKey])
-			password = string(g.secret.Data[corev1.BasicAuthPasswordKey])
-		case corev1.SecretTypeTLS:
-			cert, err := tls.X509KeyPair(g.secret.Data[corev1.TLSCertKey], g.secret.Data[corev1.TLSPrivateKeyKey])
-			if err != nil {
-				return nil, err
-			}
-			tlsConfig.Certificates = append(tlsConfig.Certificates, cert)
-		}
-	}
+// fetchCheckoutReset - fetches and checkout and then make a hard reset just for safety
+func (ro *repoOperation) fetchCheckoutReset(commit plumbing.Hash, branch string) error {
+	// Ignore the error - just fetch
+	ro.fetch(branch)
 
-	if len(g.caBundle) > 0 {
-		cert, err := x509.ParseCertificate(g.caBundle)
-		if err != nil {
-			return nil, err
-		}
-		pool, err := x509.SystemCertPool()
-		if err != nil {
-			pool = x509.NewCertPool()
-		}
-		pool.AddCert(cert)
-		tlsConfig.RootCAs = pool
+	err := ro.checkout(commit, branch)
```

Comment:
I don't think we want to checkout since checkout leaves a detached head. We need to do a reset for a clean state of the repository pointing to the branch

Reply:
Removed